ciao-deploy/deploy/utils.go

```diff
@@ -0,0 +1,317 @@
+// Copyright © 2017 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deploy
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"os/user"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
+	"golang.org/x/crypto/ssh/knownhosts"
+
+	"github.com/01org/ciao/ssntp"
+	"github.com/01org/ciao/ssntp/certs"
+	"github.com/pkg/errors"
+)
+
+// DefaultImageCacheDir provides the default location for downloaded images
+func DefaultImageCacheDir() string {
+	u, err := user.Current()
+	if err != nil {
+		return ""
+	}
+
+	return path.Join(u.HomeDir, ".cache", "ciao", "images")
+}
+
+// DownloadImage checks for a cached image in the cache directory and downloads
+// otherwise. The returned string is the path to the file and the boolean
+// indicates if it was downloaded on this function call.
+func DownloadImage(ctx context.Context, url string, imageCacheDir string) (string, bool, error) {
+	ss := strings.Split(url, "/")
+	localName := ss[len(ss)-1]
+
+	imagePath := path.Join(imageCacheDir, localName)
+	if _, err := os.Stat(imagePath); err == nil {
+		fmt.Printf("Using already downloaded image: %s\n", imagePath)
+		return imagePath, false, nil
+	} else if !os.IsNotExist(err) {
+		return "", false, errors.Wrap(err, "Error when stat()ing expected image path")
+	}
+
+	if err := os.MkdirAll(imageCacheDir, 0755); err != nil {
+		return "", false, errors.Wrap(err, "Unable to create image cache directory")
+	}
+
+	f, err := ioutil.TempFile(imageCacheDir, localName)
+	if err != nil {
+		return "", false, errors.Wrap(err, "Unable to create temporary file for download")
+	}
+	defer func() { _ = f.Close() }()
```

Comment:
As this file is opened for writing it's not safe to ignore the error from Close.